rerun_cpp/src/rerun/component_batch.hpp

```diff
@@ -1,94 +1,336 @@
+        }
+
+        /// Returns the ownership of the component batch.
+        ///
+        /// This is usually only needed for debugging and testing.
+        BatchOwnership get_ownership() const {
+            return ownership;
         }
 
       private:
-        Result<rerun::DataCell> (*to_data_cell_func)(const void*, size_t);
+        template <typename T>
+        union ComponentBatchStorage {
+            struct {
+                const T* data;
+                size_t num_instances;
+            } borrowed;
+
+            std::vector<T> vector_owned;
+
+            ComponentBatchStorage() {}
+
+            ~ComponentBatchStorage() {}
+        };
+
+        BatchOwnership ownership;
+        ComponentBatchStorage<TComponent> storage;
+    };
+
+    /// Adapter from std::vector of components.
+    ///
+    /// Only takes ownership if a temporary is passed.
+    template <typename TComponent>
+    struct ComponentBatchAdapter<TComponent, std::vector<TComponent>> {
+        ComponentBatch<TComponent> operator()(const std::vector<TComponent>& input) {
+            return ComponentBatch<TComponent>::borrow(input.data(), input.size());
+        }
+
+        ComponentBatch<TComponent> operator()(std::vector<TComponent>&& input) {
+            return ComponentBatch<TComponent>::take_ownership(std::move(input));
+        }
+    };
+
+    /// Adapter from std::array of components.
+    ///
+    /// Only takes ownership if a temporary is passed.
+    template <typename TComponent, size_t NumInstances>
+    struct ComponentBatchAdapter<TComponent, std::array<TComponent, NumInstances>> {
+        ComponentBatch<TComponent> operator()(const std::array<TComponent, NumInstances>& array) {
+            return ComponentBatch<TComponent>::borrow(array.data(), NumInstances);
+        }
+
+        ComponentBatch<TComponent> operator()(std::array<TComponent, NumInstances>&& array) {
+            return ComponentBatch<TComponent>::take_ownership(
+                std::vector<TComponent>(array.begin(), array.end())
+            );
+        }
+    };
+
+    /// Adaptor from a C-Array reference.
+    ///
+    /// *Attention*: Does *not* take ownership of the data,
+    /// you need to ensure that the data outlives the component batch.
+    template <typename TComponent, size_t NumInstances>
+    struct ComponentBatchAdapter<TComponent, TComponent[NumInstances]> {
+        ComponentBatch<TComponent> operator()(const TComponent (&array)[NumInstances]) {
+            return ComponentBatch<TComponent>::borrow(array, NumInstances);
+        }
+
+        ComponentBatch<TComponent> operator()(TComponent (&&array)[NumInstances]) {
+            std::vector<TComponent> components;
+            components.reserve(NumInstances);
+            components.insert(
+                components.end(),
+                std::make_move_iterator(array),
+                std::make_move_iterator(array + NumInstances)
```

Comment:
I think this will call the destructor of the elements twice, since you cannot leave the C-array in an empty state.

Reply:
but each element should be moved out, so that's fine. I believe.


https://en.cppreference.com/w/cpp/iterator/move_iterator
`dereferencing converts the value returned by the underlying iterator into an rvalue.`

Reply:
oh right